source/fuzz/available_instructions.cpp

```diff
@@ -0,0 +1,113 @@
+    std::vector<opt::Instruction*> params;
+    function.ForEachParam(
+        [this, &filter, ir_context, &params](opt::Instruction* param) {
+          if (filter(ir_context, param)) {
+            params.push_back(param);
+          }
+        });
+    available_params_.insert({&function, params});
+
+    auto dominator_analysis = ir_context->GetDominatorAnalysis(&function);
+    for (auto& block : function) {
+      if (&block == &*function.begin()) {
+        num_available_at_block_entry_.insert(
+            {&block, params.size() + available_globals_.size()});
+      } else {
+        auto immediate_dominator =
+            dominator_analysis->ImmediateDominator(&block);
+        num_available_at_block_entry_.insert(
+            {&block,
+             generated_by_block_.at(immediate_dominator).size() +
+                 num_available_at_block_entry_.at(immediate_dominator)});
+      }
+      std::vector<opt::Instruction*> generated;
+      for (auto& inst : block) {
+        num_available_before_instruction_.insert(
+            {&inst,
+             num_available_at_block_entry_.at(&block) + generated.size()});
+        if (filter(ir_context, &inst)) {
+          generated.push_back(&inst);
+        }
+      }
+      generated_by_block_.insert({&block, generated});
+    }
+  }
+}
+
+AvailableInstructions::AvailableBeforeInstruction::AvailableBeforeInstruction(
+    const AvailableInstructions& available_instructions, opt::Instruction* inst)
+    : available_instructions_(available_instructions), inst_(inst) {}
+
+uint32_t AvailableInstructions::AvailableBeforeInstruction::size() const {
+  return available_instructions_.num_available_before_instruction_.at(inst_);
+}
+
+bool AvailableInstructions::AvailableBeforeInstruction::empty() const {
+  return size() == 0;
+}
+
+opt::Instruction* AvailableInstructions::AvailableBeforeInstruction::operator[](
+    uint32_t index) const {
+  if (index < available_instructions_.available_globals_.size()) {
+    return available_instructions_.available_globals_[index];
+  }
+  auto block = available_instructions_.ir_context_->get_instr_block(inst_);
+  auto function = block->GetParent();
+
+  if (index <
+      available_instructions_.available_globals_.size() +
+          available_instructions_.available_params_.at(function).size()) {
+    return available_instructions_.available_params_.at(
+        function)[index - available_instructions_.available_globals_.size()];
+  }
+
+  auto dominator_analysis =
+      available_instructions_.ir_context_->GetDominatorAnalysis(function);
+
+  for (auto* ancestor = block;;
+       ancestor = dominator_analysis->ImmediateDominator(ancestor)) {
+    if (index >=
+        available_instructions_.num_available_at_block_entry_.at(ancestor)) {
+      return available_instructions_.generated_by_block_.at(
+          ancestor)[index - available_instructions_
+                                .num_available_at_block_entry_.at(ancestor)];
+    }
+  }
```

Comment:
I think in the worst case this is linear. Wouldn't that make calling this function inside a loop a quadratic time complexity (in the worst case)?

Reply:
Yes, it is - so indeed we could get quadratic complexity in the worst case. However, I don't anticipate this being too bad in practice and it's *way* more efficient than what we're currently doing. I suggest we go with it and continue profiling the fuzzer; that should tell us in due course if it becomes the dominant hot spot again (once we make everything else more efficient).